circle/src/twiddles.rs

```diff
@@ -0,0 +1,72 @@
+use alloc::vec::Vec;
+use core::mem;
+
+use itertools::Itertools;
+use p3_field::batch_multiplicative_inverse;
+use p3_field::extension::{Complex, ComplexExtendable};
+use p3_util::linear_map::LinearMap;
+use tracing::instrument;
+
+use crate::domain::twin_coset_domain;
+
+#[derive(Default)]
+pub(crate) struct TwiddleCache<F>(
+    // (log_n, shift) -> (twiddles, inverse_twiddles)
+    #[allow(clippy::type_complexity)]
+    LinearMap<(usize, Complex<F>), (Vec<Vec<F>>, Option<Vec<Vec<F>>>)>,
+);
+
+impl<F: ComplexExtendable> TwiddleCache<F> {
+    pub(crate) fn get_twiddles(
+        &mut self,
+        log_n: usize,
+        shift: Complex<F>,
+        inv: bool,
+    ) -> &Vec<Vec<F>> {
+        let cache = self
+            .0
+            .get_or_insert_with((log_n, shift), || (compute_twiddles(log_n, shift), None));
+        if !inv {
+            &cache.0
+        } else {
+            cache.1.get_or_insert_with(|| {
+                cache
+                    .0
+                    .iter()
+                    .map(|xs| batch_multiplicative_inverse(xs))
+                    .collect()
+            })
+        }
+    }
+}
+
+#[instrument(skip(shift))]
+fn compute_twiddles<F: ComplexExtendable>(log_n: usize, shift: Complex<F>) -> Vec<Vec<F>> {
+    let size = 1 << (log_n - 1);
+    let g = F::circle_two_adic_generator(log_n - 1);
+
+    let init_domain = twin_coset_domain(g, shift, size).collect_vec();
+
+    let mut working_domain: Vec<_> = init_domain
+        .iter()
+        .take(size / 2)
+        .map(|x| x.real())
+        .collect(); // After the first step we only need the real part.
```

Comment:
I think it is good to had a reference to why taking the real part for the working domain, this would be Algorithm 1 of the Circle STARKs paper, line 13.